### DISCUSSION TITLE
CS8328 question:out ref is avaliable

### DISCUSSION THREAD

**Discussion by XNTEABDSC** [dotnet/csharplang#7494] [General] [answered]
ref type is type,so we can use  "out ref T value"  to do thing like return ref

**Comment by Rekkonnect** [comment, marked as answer]
No, currently you cannot do `out ref T value`. The real type of the parameter is still `T`; `ref` is a modifier showing that the parameter is passed by reference.

To do that you need to wrap your `T` onto another object, like a `Box<T>` or something that basically acts like a reference to `T`. Your box could be either of the two, depending on how you lay your structure:

### Class
```csharp
public class Box<T>
{
    public T Value { get; set; }
}
```

### Struct - Unsafe
```csharp
public ref struct Ref<T>
{
    private ref T value;

    public ref T Value
    {
        get => ref value;
    }
    
    public unsafe void SetValue(ref T value)
    {
        this.value = ref value;
    }
}
```

Usually, the simplest and safest option is by doing the `Box<T>` class, the `Ref<T>` struct I defined above could be unsafe and harder to handle if you aren't too familiar with unsafe constructs.

**Comment by XNTEABDSC** [comment]
```
public ref T Get(int i);
public bool Has(int i);
public bool TryGet(int i,out ref T value)
```
I think my needs are reasonable.

**Reply by Rekkonnect** [reply]
In that case, a simple box would be your solution; you also avoid technicalities around using the ref you're providing, and it's clearer that you intend the ref be modified.
